File: Tut_03/Sorting/Java/MergeSort.java
package Tut_01;
import java.util.Scanner;

public class MergeSort {
	static void mergeSort(int arr[], int len) {
		if ( len > 1) {
			int mid = len / 2;
			int left[] = new int[mid];
			int right[] = new int[len - mid];
			
			int k = 0;
			for ( int i = 0; i < len; i++) {
				if ( i < mid) {
					left[i] = arr[i];
				} else {
					right[k] = arr[i];
					k = k + 1;
				}
			}
			
			mergeSort(left, mid);
			mergeSort(right, len - mid);
			
			int l = 0;
			int r = 0;
			int in = 0;
			while ( l < left.length && r < right.length) {
				if ( left[l] <= right[r]) {
					arr[in] = left[l];
					l++;
				} else {
					arr[in] = right[r];
					r++;
				}
				in++;
			}
			while(l < left.length) {
				arr[in] = left[l];
				l++;
				in++;
			}
			while(r < right.length) {
				arr[in] = right[r];
				r++;
				in++;
				
			}
			
		}
	}
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		int n = scanner.nextInt();
		int arr[] = new int[n];
		for ( int i = 0; i < n; i++) {
			System.out.printf("arr[%d] = ", i + 1);
			arr[i] = scanner.nextInt();
		}
		mergeSort(arr, n);
		for ( int i = 0; i < n; i++) {
			System.out.printf(arr[i] + " ");
		}
		scanner.close();
 	}

}

File: Tut_03/Sorting/Python/merge_sort.py
def ms(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        left = arr[mid:]
        right = arr[:mid]

        ms(left)
        ms(right)
        l = 0
        r = 0
        i = 0
        while l < len(left) and r < len(right):
            if left[l] < right[r]:
                arr[i] = left[l]
                l += 1
            else:
                arr[i] = right[r]
                r += 1
            i += 1
        while l < len(left):
            arr[i] = left[l]
            i += 1
            l += 1
        while r < len(right):
            arr[i] = right[r]
            i += 1
            r += 1


if __name__ == '__main__':
    arr = [3, 2, 5, 2, 5, 1, 5, 2, 43, 2, 3, 5, 21, 4]
    ms(arr)
    print(arr)

